File: reference/src/policy/expression.ts
/**
 * Policy expression evaluator.
 * Provides a simple expression language for gate rule evaluation.
 * For Rego/CEL, users implement the ExpressionEvaluator interface.
 */

export interface ExpressionEvaluator {
  /** Evaluate an expression against a context. Returns true/false. */
  evaluate(expression: string, context: Record<string, unknown>): boolean;
  /** Optionally validate an expression before runtime. */
  validate?(expression: string): { valid: boolean; error?: string };
}

import type { ExpressionRule } from '../core/types.js';
export type { ExpressionRule } from '../core/types.js';

export interface ExpressionVerdict {
  passed: boolean;
  message?: string;
}

/**
 * Resolve a dotted property path on an object.
 * E.g., "ctx.metrics.coverage" resolves context.metrics.coverage.
 */
function resolveProperty(path: string, context: Record<string, unknown>): unknown {
  const parts = path.split('.');
  let current: unknown = context;
  for (const part of parts) {
    if (current === null || current === undefined || typeof current !== 'object') {
      return undefined;
    }
    current = (current as Record<string, unknown>)[part];
  }
  return current;
}

function toNumber(value: unknown): number | undefined {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const n = Number(value);
    return isNaN(n) ? undefined : n;
  }
  return undefined;
}

function resolveValue(token: string, context: Record<string, unknown>): unknown {
  // String literal
  if (
    (token.startsWith("'") && token.endsWith("'")) ||
    (token.startsWith('"') && token.endsWith('"'))
  ) {
    return token.slice(1, -1);
  }
  // Boolean literals
  if (token === 'true') return true;
  if (token === 'false') return false;
  // Number literal
  const num = Number(token);
  if (!isNaN(num) && token.trim() !== '') return num;
  // Property path
  return resolveProperty(token, context);
}

/**
 * Evaluate a simple comparison expression.
 * Supports: >=, <=, ==, !=, >, <
 */
function evaluateComparison(
  left: string,
  operator: string,
  right: string,
  context: Record<string, unknown>,
): boolean {
  const lVal = resolveValue(left.trim(), context);
  const rVal = resolveValue(right.trim(), context);

  const lNum = toNumber(lVal);
  const rNum = toNumber(rVal);

  switch (operator) {
    case '>=':
      return lNum !== undefined && rNum !== undefined && lNum >= rNum;
    case '<=':
      return lNum !== undefined && rNum !== undefined && lNum <= rNum;
    case '>':
      return lNum !== undefined && rNum !== undefined && lNum > rNum;
    case '<':
      return lNum !== undefined && rNum !== undefined && lNum < rNum;
    case '==':
      return lVal === rVal || (lNum !== undefined && rNum !== undefined && lNum === rNum);
    case '!=':
      if (lNum !== undefined && rNum !== undefined) return lNum !== rNum;
      return lVal !== rVal;
    default:
      return false;
  }
}

/**
 * Evaluate a "contains" expression: `collection contains value`
 */
function evaluateContains(
  collectionPath: string,
  valuePath: string,
  context: Record<string, unknown>,
): boolean {
  const collection = resolveValue(collectionPath.trim(), context);
  const value = resolveValue(valuePath.trim(), context);
  if (Array.isArray(collection)) {
    return collection.includes(value);
  }
  if (typeof collection === 'string' && typeof value === 'string') {
    return collection.includes(value);
  }
  return false;
}

/**
 * Evaluate a single atomic expression (no logical operators).
 */
function evaluateAtomic(expression: string, context: Record<string, unknown>): boolean {
  const trimmed = expression.trim();

  // Negation
  if (trimmed.startsWith('!')) {
    return !evaluateAtomic(trimmed.slice(1), context);
  }

  // Contains
  const containsIdx = trimmed.indexOf(' contains ');
  if (containsIdx !== -1) {
    return evaluateContains(
      trimmed.slice(0, containsIdx),
      trimmed.slice(containsIdx + ' contains '.length),
      context,
    );
  }

  // Comparison operators (longest first to avoid partial matches)
  for (const op of ['>=', '<=', '!=', '==', '>', '<']) {
    const opIdx = trimmed.indexOf(op);
    if (opIdx !== -1) {
      return evaluateComparison(
        trimmed.slice(0, opIdx),
        op,
        trimmed.slice(opIdx + op.length),
        context,
      );
    }
  }

  // Truthiness check
  const val = resolveValue(trimmed, context);
  return Boolean(val);
}

/**
 * Create a simple expression evaluator.
 * Handles comparisons (>=, <=, ==, !=, >, <), logical (&&, ||, !),
 * property access (ctx.metrics.coverage), and set membership (contains).
 */
export function createSimpleExpressionEvaluator(): ExpressionEvaluator {
  return {
    evaluate(expression: string, context: Record<string, unknown>): boolean {
      // Handle || (OR) — lowest precedence
      if (expression.includes('||')) {
        const parts = expression.split('||');
        return parts.some((part) => this.evaluate(part.trim(), context));
      }

      // Handle && (AND)
      if (expression.includes('&&')) {
        const parts = expression.split('&&');
        return parts.every((part) => this.evaluate(part.trim(), context));
      }

      return evaluateAtomic(expression, context);
    },

    validate(expression: string): { valid: boolean; error?: string } {
      if (!expression || expression.trim().length === 0) {
        return { valid: false, error: 'Expression is empty' };
      }
      // Basic validation: check for balanced structure
      const hasOperator =
        expression.includes('>=') ||
        expression.includes('<=') ||
        expression.includes('==') ||
        expression.includes('!=') ||
        expression.includes('>') ||
        expression.includes('<') ||
        expression.includes('contains') ||
        expression.includes('&&') ||
        expression.includes('||') ||
        expression.startsWith('!');

      if (!hasOperator) {
        // Could be a simple truthiness check — still valid
        return { valid: true };
      }
      return { valid: true };
    },
  };
}

/**
 * Evaluate an expression rule and return a gate verdict.
 */
export function evaluateExpressionRule(
  rule: ExpressionRule,
  context: Record<string, unknown>,
  evaluator: ExpressionEvaluator,
): ExpressionVerdict {
  try {
    const passed = evaluator.evaluate(rule.expression, context);
    return {
      passed,
      message: passed ? undefined : `Expression failed: ${rule.expression}`,
    };
  } catch (err) {
    return {
      passed: false,
      message: `Expression error: ${err instanceof Error ? err.message : String(err)}`,
    };
  }
}

File: sdk-python/src/ai_sdlc/policy/expression.py
"""Policy expression evaluator.

Provides a simple expression language for gate rule evaluation.
For Rego/CEL, users implement the ExpressionEvaluator protocol.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Protocol

if TYPE_CHECKING:
    from ai_sdlc.core.types import ExpressionRule


@dataclass(frozen=True)
class ExpressionVerdict:
    passed: bool
    message: str | None = None


class ExpressionEvaluator(Protocol):
    """Evaluate an expression against a context. Returns true/false."""

    def evaluate(self, expression: str, context: dict[str, Any]) -> bool: ...

    def validate(self, expression: str) -> dict[str, Any]:
        """Optionally validate an expression before runtime."""
        ...


def _resolve_property(path: str, context: dict[str, Any]) -> Any:
    """Resolve a dotted property path on a dict."""
    parts = path.split(".")
    current: Any = context
    for part in parts:
        if current is None or not isinstance(current, dict):
            return None
        current = current.get(part)
    return current


def _to_number(value: Any) -> float | None:
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    return None


def _resolve_value(token: str, context: dict[str, Any]) -> Any:
    token = token.strip()
    # String literal
    if (token.startswith("'") and token.endswith("'")) or (
        token.startswith('"') and token.endswith('"')
    ):
        return token[1:-1]
    # Boolean literals
    if token == "true":
        return True
    if token == "false":
        return False
    # Number literal
    try:
        return float(token) if "." in token else int(token)
    except ValueError:
        pass
    # Property path
    return _resolve_property(token, context)


def _evaluate_comparison(
    left: str, operator: str, right: str, context: dict[str, Any]
) -> bool:
    l_val = _resolve_value(left.strip(), context)
    r_val = _resolve_value(right.strip(), context)

    l_num = _to_number(l_val)
    r_num = _to_number(r_val)

    if operator == ">=":
        return l_num is not None and r_num is not None and l_num >= r_num
    if operator == "<=":
        return l_num is not None and r_num is not None and l_num <= r_num
    if operator == ">":
        return l_num is not None and r_num is not None and l_num > r_num
    if operator == "<":
        return l_num is not None and r_num is not None and l_num < r_num
    if operator == "==":
        if l_val == r_val:
            return True
        return l_num is not None and r_num is not None and l_num == r_num
    if operator == "!=":
        if l_num is not None and r_num is not None:
            return bool(l_num != r_num)
        return bool(l_val != r_val)
    return False


def _evaluate_contains(
    collection_path: str, value_path: str, context: dict[str, Any]
) -> bool:
    collection = _resolve_value(collection_path.strip(), context)
    value = _resolve_value(value_path.strip(), context)
    if isinstance(collection, list):
        return value in collection
    if isinstance(collection, str) and isinstance(value, str):
        return value in collection
    return False


def _evaluate_atomic(expression: str, context: dict[str, Any]) -> bool:
    trimmed = expression.strip()

    # Negation
    if trimmed.startswith("!"):
        return not _evaluate_atomic(trimmed[1:], context)

    # Contains
    idx = trimmed.find(" contains ")
    if idx != -1:
        return _evaluate_contains(
            trimmed[:idx], trimmed[idx + len(" contains "):], context
        )

    # Comparison operators (longest first)
    for op in (">=", "<=", "!=", "==", ">", "<"):
        op_idx = trimmed.find(op)
        if op_idx != -1:
            return _evaluate_comparison(
                trimmed[:op_idx], op, trimmed[op_idx + len(op):], context
            )

    # Truthiness check
    val = _resolve_value(trimmed, context)
    return bool(val)


class _SimpleExpressionEvaluator:
    def evaluate(self, expression: str, context: dict[str, Any]) -> bool:
        # Handle || (OR) - lowest precedence
        if "||" in expression:
            parts = expression.split("||")
            return any(self.evaluate(part.strip(), context) for part in parts)

        # Handle && (AND)
        if "&&" in expression:
            parts = expression.split("&&")
            return all(self.evaluate(part.strip(), context) for part in parts)

        return _evaluate_atomic(expression, context)

    def validate(self, expression: str) -> dict[str, Any]:
        if not expression or expression.strip() == "":
            return {"valid": False, "error": "Expression is empty"}
        return {"valid": True}


def create_simple_expression_evaluator() -> ExpressionEvaluator:
    """Create a simple expression evaluator.

    Handles comparisons (>=, <=, ==, !=, >, <), logical (&&, ||, !),
    property access (ctx.metrics.coverage), and set membership (contains).
    """
    return _SimpleExpressionEvaluator()


def evaluate_expression_rule(
    rule: ExpressionRule,
    context: dict[str, Any],
    evaluator: ExpressionEvaluator,
) -> ExpressionVerdict:
    """Evaluate an expression rule and return a gate verdict."""
    try:
        passed = evaluator.evaluate(rule.expression, context)
        return ExpressionVerdict(
            passed=passed,
            message=None if passed else f"Expression failed: {rule.expression}",
        )
    except Exception as err:
        return ExpressionVerdict(
            passed=False,
            message=f"Expression error: {err}",
        )
